libs/compute_api/src/spec.rs

```diff
@@ -3,37 +3,190 @@
 //! The spec.json file is used to pass information to 'compute_ctl'. It contains
 //! all the information needed to start up the right version of PostgreSQL,
 //! and connect it to the storage nodes.
-use serde::Deserialize;
+use anyhow::anyhow;
+use serde::{Deserialize, Serialize};
+use serde_with::{serde_as, DisplayFromStr};
 use std::collections::HashMap;
+use std::str::FromStr;
+use utils::id::{TenantId, TimelineId};
+use utils::lsn::Lsn;
 
 /// String type alias representing Postgres identifier and
 /// intended to be used for DB / role names.
 pub type PgIdent = String;
 
 /// Cluster spec or configuration represented as an optional number of
 /// delta operations + final cluster state description.
-#[derive(Clone, Deserialize)]
-pub struct ComputeSpec {
-    pub format_version: f32,
-    pub timestamp: String,
+#[serde_as]
+#[derive(Clone, Deserialize, Serialize)]
+pub struct ComputeSpecV2 {
+    pub format_version: u64,
+
+    // For debugging purposes only
+    pub project_id: Option<String>,
+    pub endpoint_id: Option<String>,
     pub operation_uuid: Option<String>,
-    /// Expected cluster state at the end of transition process.
-    pub cluster: Cluster,
+
+    /// W3C trace context of the launch operation, for OpenTelemetry tracing
+    pub startup_tracing_context: Option<HashMap<String, String>>,
+
+    // Information needed to connect to the storage layer.
+    //
+    // `tenant_id`, `timeline_id` and `pageserver_connstring` are always needed.
+    //
+    // If Lsn == None, this is a primary endpoint that continues writing WAL at
+    // the end of the timeline. If 'lsn' is set, this is a read-only node
+    // "anchored" at that LSN. 'safekeeper_connstrings' must be non-empty for a
+    // primary.
+    #[serde_as(as = "DisplayFromStr")]
+    pub tenant_id: TenantId,
+    #[serde_as(as = "DisplayFromStr")]
+    pub timeline_id: TimelineId,
+    #[serde_as(as = "Option<DisplayFromStr>")]
+    pub lsn: Option<Lsn>,
+    pub pageserver_connstring: String,
+    pub safekeeper_connstrings: Vec<String>,
+
+    /// If set, 'storage_auth_token' is used as the password to authenticate to
+    /// the pageserver and safekeepers.
+    pub storage_auth_token: Option<String>,
+
+    /// Contents of postgresql.conf file
+    pub postgresql_conf: Option<String>,
+
+    /// Extra settings to append to the postgresql.conf
+    pub settings: GenericOptions,
+
+    // Expected cluster state at the end of transition process.
+    pub roles: Vec<Role>,
+    pub databases: Vec<Database>,
+    pub extensions: Vec<PgIdent>,
     pub delta_operations: Option<Vec<DeltaOp>>,
```

Comment:
> Is this still the same DeltaOp with stringy action identifiers?

Yes.

> If so, can we update that to typed action operations?

Good idea

Reply:
I extracted these spec format changes to https://github.com/neondatabase/neon/pull/4007 since the first version of this PR, so this comment now applies to that PR, not this one. I'm marking this as resolved, and copying the comment to that.